src/tests/ftest/control/dmg_nvme_scan_test.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2018-2019 Intel Corporation.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+  The Government's rights to use, modify, reproduce, release, perform, display,
+  or disclose this software are subject to the terms of the Apache License as
+  provided in Contract No. B609815.
+  Any reproduction of computer software, computer software documentation, or
+  portions thereof marked with this legend must also reproduce the markings.
+"""
+from __future__ import print_function
+
+from dmg_utils import DmgCommand
+from server_utils import ServerCommand
+from apricot import TestWithoutServers
+from avocado.utils import process
+
+class DmgNvmeScanTest(TestWithoutServers):
+    """Test Class Description:
+    Simple test to verify the scan function of the dmg tool.
+    :avocado: recursive
+    """
+
+    def test_dmg_nvme_scan_basic(self):
+        """
+        JIRA ID: DAOS-2485
+        Test Description: Test basic dmg functionality to scan nvme the storage
+        on system.
+        :avocado: tags=all,hw,dmg,control
+        """
+        # Create daos_server command
+        server = ServerCommand(self.hostlist_servers)
+        server.get_params(self, "/run/daos_server/*")
+
+        # Update config and start server
+        server.update_configuration(self.basepath)
+        server.prepare(self.workdir, self.hostfile_servers_slots)
+        server.start(self.orterun)
+
+        # Create daos_shell command
+        dmg = DmgCommand()
+        dmg.get_params(self, "/run/dmg/*")
+
+        # Update hostlist value for dmg command
+        ports = self.params.get("ports", "/run/hosts/*")
+
+        # Check that hosts and ports are same length
+        self.assertEqual(ports, self.hostlist_servers)
+
+        servers_with_ports = ["{}:{}".format(host, ports[i])
+                              for i, host in enumerate(self.hostlist_servers)]
```

Comment:
The `TestWithoutServers` code does not assign the `self.hostlist_servers` attribute, so code similar to what exists in `TestWithServers.setup()` is needed.  One option would be to inherit from `TestWithServers`, but I'm not sure you need to be starting agents.  If not then it would make sense to separate out the processing and assigning of host lists into a method defined in `TestWithoutServers`, that `TestWithServers` would call in its `setup()` method, specifically this code:

```
        # Determine which hosts to use as servers and optionally clients.
        # Support the use of a host type count to test with subsets of the
        # specified hosts lists
        test_machines = self.params.get("test_machines", "/run/hosts/*")
        test_servers = self.params.get("test_servers", "/run/hosts/*")
        test_clients = self.params.get("test_clients", "/run/hosts/*")
        server_count = self.params.get("server_count", "/run/hosts/*")
        client_count = self.params.get("client_count", "/run/hosts/*")

        # Supported combinations of yaml hosts arguments:
        #   - test_machines [+ server_count]
        #   - test_servers [+ server_count]
        #   - test_servers [+ server_count] + test_clients [+ client_count]
        if test_machines:
            self.hostlist_servers = test_machines[:server_count]
        elif test_servers and test_clients:
            self.hostlist_servers = test_servers[:server_count]
            self.hostlist_clients = test_clients[:client_count]
        elif test_servers:
            self.hostlist_servers = test_servers[:server_count]
        self.log.info("hostlist_servers:  %s", self.hostlist_servers)
        self.log.info("hostlist_clients:  %s", self.hostlist_clients)

        # If a specific count is specified, verify enough servers/clients are
        # specified to satisy the count
        host_count_checks = (
            ("server", server_count,
             len(self.hostlist_servers) if self.hostlist_servers else 0),
            ("client", client_count,
             len(self.hostlist_clients) if self.hostlist_clients else 0)
        )
        for host_type, expected_count, actual_count in host_count_checks:
            if expected_count:
                self.assertEqual(
                    expected_count, actual_count,
                    "Test requires {} {}; {} specified".format(
                        expected_count, host_type, actual_count))
```

Reply:
The `DmgNvmeScanTest` still needs to call `self.get_test_attributes()` to define `self.hostlist_servers` either in an overridden `setUp()` method or fairly early on in the `test_dmg_nvme_scan_basic()` method.

Reply:
I included a setUp on the test calling get_test_attributes().